graylog2-server/src/main/java/org/graylog2/migrations/V20161125161400_AlertReceiversMigration.java

```diff
@@ -0,0 +1,172 @@
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.migrations;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.auto.value.AutoValue;
+import com.mongodb.BasicDBObject;
+import com.mongodb.DBCollection;
+import org.bson.types.ObjectId;
+import org.graylog2.alarmcallbacks.AlarmCallbackConfiguration;
+import org.graylog2.alarmcallbacks.AlarmCallbackConfigurationAVImpl;
+import org.graylog2.alarmcallbacks.AlarmCallbackConfigurationService;
+import org.graylog2.alarmcallbacks.EmailAlarmCallback;
+import org.graylog2.database.CollectionName;
+import org.graylog2.database.MongoConnection;
+import org.graylog2.plugin.cluster.ClusterConfigService;
+import org.graylog2.plugin.database.Persisted;
+import org.graylog2.plugin.database.ValidationException;
+import org.graylog2.streams.StreamImpl;
+import org.graylog2.streams.StreamService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class V20161125161400_AlertReceiversMigration extends Migration {
+    private static final Logger LOG = LoggerFactory.getLogger(V20161125161400_AlertReceiversMigration.class);
+
+    private final ClusterConfigService clusterConfigService;
+    private final StreamService streamService;
+    private final AlarmCallbackConfigurationService alarmCallbackService;
+    private final DBCollection dbCollection;
+
+    @Inject
+    public V20161125161400_AlertReceiversMigration(ClusterConfigService clusterConfigService,
+                                                   StreamService streamService,
+                                                   AlarmCallbackConfigurationService alarmCallbackService,
+                                                   MongoConnection mongoConnection) {
+        this.clusterConfigService = clusterConfigService;
+        this.streamService = streamService;
+        this.alarmCallbackService = alarmCallbackService;
+        final String collectionName = StreamImpl.class.getAnnotation(CollectionName.class).value();
+        this.dbCollection = mongoConnection.getDatabase().getCollection(collectionName);
+    }
+
+    @Override
+    public ZonedDateTime createdAt() {
+        return ZonedDateTime.parse("2016-11-25T16:14:00Z");
+    }
+
+    @Override
+    public void upgrade() {
+        // Do not run again if the migration marker can be found in the database.
+        if (clusterConfigService.get(MigrationCompleted.class) != null) {
+            return;
+        }
+
+        final Map<String, Optional<String>> streamMigrations = this.streamService.loadAll()
+                .stream()
+                .filter(stream -> !stream.getAlertReceivers().isEmpty()
```

Comment:
This is wrong. It only checks if there are no keys in the map but not if the values for the keys are empty.